### ISSUE TITLE
Doxygen 1.8.17: regression of C macro expansion

### ISSUE THREAD

**Issue by fgrisez** [doxygen/doxygen#7527]
Hi,

I think release of Doxygen 1.8.17 has introduced a regression in its C macro preprocessor.

Given the following C header file:
```C
/**
 * @file c-types.h
 */

#define L_CONCAT__(A, B) A ## B
#define L_CONCAT_(A, B) L_CONCAT__(A, B)
#define L_CONCAT(A, B) L_CONCAT_(A, B)

#define L_C_ENUM_PREFIX MyProject

#define L_DECLARE_C_ENUM(NAME) \
    typedef enum L_CONCAT(_, L_CONCAT(L_C_ENUM_PREFIX, NAME)) { \
        Enumerator1, \
        Enumerator2, \
        Enumerator3 \
    } L_CONCAT(L_C_ENUM_PREFIX, NAME)

/**
 * @brief MyEnum
 */
L_DECLARE_C_ENUM(Enum);
```
Doxygen settings:

+ ENABLE_PREPROCESSING = YES
+ MACRO_EXPANSION = YES
+ GENERATE_XML = YES

Whatever the version of Doxygen, the enum section is correctly generated but the typedef section is incorrect with Doxygen 1.8.17. See following snippet:

```xml
<memberdef kind="typedef" id="c-types_8h_1a930ef04dea7f583542eaa9d2f3f81bf8" prot="public" static="no">
        <type>enum <ref refid="c-types_8h_1a3046793ad13c57ab53727e1050c9a8cc" kindref="member">_MyProjectEnum</ref></type>
        <definition>typedef enum _MyProjectEnum L_C_ENUM_PREFIXEnum</definition>
        <argsstring></argsstring>
        <name>L_C_ENUM_PREFIXEnum</name>
        <briefdescription>
<para>MyEnum. </para>
[…]
</memberdef>
```
whereas L_C_ENUM_PREFIX macro was well expanded with earlier version of Doxygen.

**Comment by albert-github** [comment]
There have been a number of changes regarding the preprocessor and in between these changes there have been some versions that did work and some that didn't.

The last version (as far as I can see) that did work was

> Commit: 5fe991236c9042a1ea75d6693aca41b3256a0d89 [5fe9912]
> Parents: 73a410a00e
> Date: Monday, November 25, 2019 9:02:53 PM
> Fixed preprocessor issue that resulted in "More #endif's than #if's found." warning.

the next version:

> Commit: dbd8ceb8efa2e7b413780a43e74b30f039670a72 [dbd8ceb]
> Parents: 5fe991236c
> Date: Monday, November 25, 2019 9:29:51 PM
> Fixed preprocessor hangup regression

showed the mentioned problem i.e. showed `L_C_ENUM_PREFIXEnum` instead of `MyProjectEnum`.

**Comment by disa6302** [comment]
Has this been released yet?

**Comment by albert-github** [comment]
@disa6302 
As you can see the label "fixed but not released" is still present indicating that this is not yet in any official release, it is just available in the current master version. 
When a new doxygen release becomes available (not time set yet) , the issue will be closed and the label "fixed but not released" will be removed.

**Comment by disa6302** [comment]
Sure. Thanks!

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.18.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
